In [1]:
%matplotlib inline
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import pandas as pd
from IPython.display import display
from IPython.display import Image
sns.set_context("notebook")

# Extracting Data and Filtering
:The `[]` operator supports array of booleans as an input. In this case, the array must be exactly as long as the number of rows.

The result is a **filtered version of the data frame**, where **only rows corresponding to True appear**.

# Q1.
Select multiple criteria by creating multiple boolean Series and combining them using the `&` operator. For Examole Result is win and %win is more than 50.

In [2]:
 # reading a CSV file into a data frame
df=pd.read_csv('./data/elections.csv')

 #Note: if we end a cell with an expression or variable
 #  name, the result will print the data frame
df

Year          Candidate                  Party  Popular vote Result  \
0    1824     Andrew Jackson  Democratic-Republican        151271   loss   
1    1824  John Quincy Adams  Democratic-Republican        113142    win   
2    1828     Andrew Jackson             Democratic        642806    win   
3    1828  John Quincy Adams    National Republican        500897   loss   
4    1832     Andrew Jackson             Democratic        702735    win   
..    ...                ...                    ...           ...    ...   
177  2016         Jill Stein                  Green       1457226   loss   
178  2020       Joseph Biden             Democratic      81268924    win   
179  2020       Donald Trump             Republican      74216154   loss   
180  2020       Jo Jorgensen            Libertarian       1865724   loss   
181  2020     Howard Hawkins                  Green        405035   loss   

             %  
0    57.210122  
1    42.789878  
2    56.203927  
3    43.796073  
4    54.574789  
..         ...  
177   1.073699  
178  51.311515  
179  46.858542  
180   1.177979  
181   0.255731  

[182 rows x 6 columns]

In [3]:
'''apply a boolean condition using the appropriate operator on 2 different conditions 
like Result should be WIN and % should be GREATER THAN 50'''
df_win=df[(df['Result']=='win')&(df['%'] > 50)].head()
df_win

Year               Candidate       Party  Popular vote Result          %
2   1828          Andrew Jackson  Democratic        642806    win  56.203927
4   1832          Andrew Jackson  Democratic        702735    win  54.574789
8   1836        Martin Van Buren  Democratic        763291    win  52.272472
11  1840  William Henry Harrison        Whig       1275583    win  53.051213
13  1844              James Polk  Democratic       1339570    win  50.749477

We will nowe work with a dataset containing the names of all babies born in California, USA. Visit the [Social Security Adminisitration Website](https://www.ssa.gov/oact/babynames/limits.html) for more details about the dataset if you want.

**Let's start by loading the Baby Names dataset.**

In [4]:
import urllib.request
import os.path
import zipfile
import pandas as pd

data_url = "https://www.ssa.gov/oact/babynames/state/namesbystate.zip"
local_filename = "babynamesbystate.zip"
if not os.path.exists(local_filename): # if the data exists don't download again
    with urllib.request.urlopen(data_url) as resp, open(local_filename, 'wb') as f:
        f.write(resp.read())

zf = zipfile.ZipFile(local_filename, 'r')

ca_name = 'CA.TXT'
field_names = ['State', 'Sex', 'Year', 'Name', 'Count']
with zf.open(ca_name) as fh:
    babynames = pd.read_csv(fh, header=None, names=field_names)

babynames.head(15)

State Sex  Year       Name  Count
0     CA   F  1910       Mary    295
1     CA   F  1910      Helen    239
2     CA   F  1910    Dorothy    220
3     CA   F  1910   Margaret    163
4     CA   F  1910    Frances    134
5     CA   F  1910       Ruth    128
6     CA   F  1910     Evelyn    126
7     CA   F  1910      Alice    118
8     CA   F  1910   Virginia    101
9     CA   F  1910  Elizabeth     93
10    CA   F  1910   Florence     93
11    CA   F  1910      Marie     90
12    CA   F  1910    Mildred     90
13    CA   F  1910       Rose     74
14    CA   F  1910      Hazel     68

In [5]:
#display size
babynames.size

2069470

# Q2.
Calculate the some statics of babynames DataFrame using Utility Functions, like .describe(), .size, .shape and report your analysis in comments

In [6]:
#code here
babynames.describe()

Year          Count
count  413894.000000  413894.000000
mean     1986.314353      79.140348
std        27.190404     292.028292
min      1910.000000       5.000000
25%      1969.000000       7.000000
50%      1992.000000      13.000000
75%      2008.000000      38.000000
max      2023.000000    8263.000000

In [7]:
#code here
babynames.shape

(413894, 5)

In [8]:
babynames.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 413894 entries, 0 to 413893
Data columns (total 5 columns):
 #   Column  Non-Null Count   Dtype 
---  ------  --------------   ----- 
 0   State   413894 non-null  object
 1   Sex     413894 non-null  object
 2   Year    413894 non-null  int64 
 3   Name    413894 non-null  object
 4   Count   413894 non-null  int64 
dtypes: int64(2), object(3)
memory usage: 15.8+ MB


**Add Your comments Here**
1. The data frame has `413894` entries and `5` features.
2. The data set has `2069470` total number of elements.
3. The first year of the data is `1910` having count of `5`.
4. The last year of the data is `2023` having count of `8263`.
5. There are 2 integer columns named as `Year` and `Count`.
6. There are 3 object columns which are `State` , `Sex` and `Name`.
7. There is no null values in the dataset. 

# Q.3:

Using the babynames DataFrame that contains baby names data for the state of California, how can you use the 'isin' method to find out which records have names that are in a given list of names, such as "Emma," "Liam," and "Olivia," and what would be the output of this operation when applied to the first 15 rows of the DataFrame (babynames.head(15))?

In [9]:
# List of names to check
name_list=['Emma','Liam','Olivia']
# Use the isin method to filter records with names in the name_list
filtered_names=babynames[babynames['Name'].isin(name_list)]
# Display the filtered DataFrame with appropriate head value i.e., 15
filtered_names.head(10)

State Sex  Year    Name  Count
53      CA   F  1910    Emma     29
293     CA   F  1911    Emma     29
525     CA   F  1912    Emma     56
701     CA   F  1912  Olivia      8
840     CA   F  1913    Emma     56
1168    CA   F  1914    Emma     65
1437    CA   F  1914  Olivia      6
1542    CA   F  1915    Emma     86
1869    CA   F  1915  Olivia      6
2009    CA   F  1916    Emma     78

# Q.4:
How can you use the 'str.startswith' method in pandas to filter records in a babyname DataFrame? Provide an example code using the babynames DataFrame and explain how this method can be employed to extract data for baby names that begin with 'A' or 'B' set of letters.

In [10]:
# Filter rows where the 'Name' column starts with 'A' OR 'B'
filtered_rows=babynames[babynames['Name'].str.startswith(('A','B'))]
# Display the filtered DataFrame
filtered_rows
#print("Baby names starting with 'A' and B:")
print("Baby names starting with 'A' and B:",filtered_rows)

Baby names starting with 'A' and B:        State Sex  Year      Name  Count
7         CA   F  1910     Alice    118
23        CA   F  1910   Bernice     59
31        CA   F  1910      Anna     51
32        CA   F  1910       Ann     47
36        CA   F  1910     Betty     40
...      ...  ..   ...       ...    ...
413597    CA   M  2023  Bradford      5
413598    CA   M  2023    Brahim      5
413599    CA   M  2023   Braiden      5
413600    CA   M  2023      Bret      5
413601    CA   M  2023  Brighton      5

[62932 rows x 5 columns]


# Q.5

Find the most popular baby name in California in 2012

__Exercise__: Try to find the most popular baby name by finding the maximum value for 'Count' and then apply filtering to find the name.


In [11]:
# Step 5: Find the maximum count and the corresponding name(s)
babyname_2012=babynames[babynames['Year']==2012]
max_count=babyname_2012['Count'].max()
#Hint: You should apply max on the count but try to get the whole data of that maximum row rather than try to get only the max count.
count_name=babyname_2012[babyname_2012['Count']==max_count]
count_name

State Sex  Year    Name  Count
197543    CA   F  2012  Sophia   3645

# Q. 6
Find baby names that start with J, in the Year from 2001 to 2019

*Note: To answer this question, we need a way to look at the individual characters of the baby names*

In [12]:
# Note: Series has powerful string manipulation methods associated with it
filtered_data=babynames[(babynames['Year']>=2001)&
                      (babynames['Year']<=2019)&
                      babynames['Name'].str.startswith('J')
                      ]
# Step 5: Find names that start with 'J'
start_j=filtered_data.head(15)
#Hint: 3 different conditions in a single boolean condition can be used using an operator

# Step 6: Display the filtered DataFrame with first 15 rows
start_j

State Sex  Year        Name  Count
152833    CA   F  2001     Jessica   2244
152835    CA   F  2001    Jennifer   2027
152839    CA   F  2001     Jasmine   1655
152854    CA   F  2001  Jacqueline   1195
152864    CA   F  2001       Julia    983
152873    CA   F  2001     Jocelyn    856
152920    CA   F  2001      Jordan    510
152925    CA   F  2001      Jasmin    500
152929    CA   F  2001     Juliana    494
152931    CA   F  2001       Jenna    490
152938    CA   F  2001      Jazmin    465
152952    CA   F  2001      Joanna    396
152955    CA   F  2001        Jade    384
152969    CA   F  2001     Jazmine    346
152997    CA   F  2001    Julianna    282

# Q.7
Add a new column called "name_length" to the babynames DataFrame, you can use .str.len() in existing Name column. Moreover. also modify the Length values with new values like existing_value-1 and rename the length column with new label as desceribe in slides.



In [13]:
# Calculate the length of each name and add it as a new column "_length"
babynames['name_length'] =babynames["Name"].str.len()
# Modify 'name_length' values to be existing_value - 1
babynames['name_length']=babynames['Name'].str.len()-1
# Rename the 'name_length' column to the new label 'name_length_modified'
babynames.rename(columns={'name_length':'Length'},inplace=True)
babynames

State Sex  Year      Name  Count  Length
0         CA   F  1910      Mary    295       3
1         CA   F  1910     Helen    239       4
2         CA   F  1910   Dorothy    220       6
3         CA   F  1910  Margaret    163       7
4         CA   F  1910   Frances    134       6
...      ...  ..   ...       ...    ...     ...
413889    CA   M  2023      Ziah      5       3
413890    CA   M  2023    Ziaire      5       5
413891    CA   M  2023    Zidane      5       5
413892    CA   M  2023      Zyan      5       3
413893    CA   M  2023      Zyir      5       3

[413894 rows x 6 columns]

# Q.8

Use .str.split() to sample the names with 'ad' string and show first 10 samples.

In [14]:
# Step 4: Filter names that contain 'ad' using str.contains and store it in the variable named ad_names
ad_names = babynames[babynames['Name'].str.contains('ad')]
# Step 5: Use .str.split() to split names containing 'ad' in ad_names (for demonstration purposes)
ad_names['Name_split'] = ad_names['Name'].str.split('ad')

# Display the filtered and split names
ad_names[['Name', 'Name_split']].head()

C:\Users\nasrullah\AppData\Local\Temp\ipykernel_1712\3190536009.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  ad_names['Name_split'] = ad_names['Name'].str.split('ad')


Name  Name_split
19     Gladys    [Gl, ys]
96   Madeline  [M, eline]
136     Sadie     [S, ie]
257    Gladys    [Gl, ys]
346    Nadine    [N, ine]

More on str methods when we'll study regular expressions in this course. Note that there are other approaches that we can use to find the same answer as above.

# Q.9
 Sort names by their length using lambda function.

In [15]:
#Hint: focus on 2 columns here i.e., NAME and LENGTH (apply appropriate parameters using lambda)
babynames_sorted = babynames.sort_values(by='Length', key=lambda x: x,ascending=False)
babynames_sorted[['Name','Length','Year']]

Name  Length  Year
341130  Ryanchristopher      14  1997
348093  Franciscojavier      14  2000
331124  Ryanchristopher      14  1993
337819  Franciscojavier      14  1996
317627  Franciscojavier      14  1988
...                 ...     ...   ...
114679               An       1  1990
352991               Jc       1  2002
244981               Al       1  1916
118952               Vi       1  1991
325806               Cj       1  1991

[413894 rows x 3 columns]

# Q.10
1: The code randomly mixes up the list of baby names. Why might someone want to do this, and how could it affect the analysis of the names?

2: After mixing, the code figures out how long each baby name is. Why would we care about name lengths, and what can this tell us about the names?

3: The code arranges the name lengths from longest to shortest. Why might we want to do this, and what could it help us find in the data?

4: The code finds the order of baby names based on their lengths and shows us the first few names in this new order. When might we need to organize names this way, and what might we discover from the first few names in this order?

In [16]:
#This is an entirely theoretical question and you are required to explain your logic in your own words. Make sure not to plagarise your content.

#Part1, shuffling can be done using sample(frac=1)
'''
1. Mixing up names does not effect the results.
2. It gives the fair view of all the names.
'''
#Part2
'''
1. The length of baby name is ranges from 1 to 14.
2. Because we could see that the people of state likes short names or long names.
'''
#Part3
'''
1. Because we can find the longest names and shortest names.
2. It identifies popular names of the State.
'''

#Part4
'''
1. Because We can see quickely the unique names by length.
2. We see the longer name which is most common in State.
'''


'\n1. Because We can see quickely the unique names by length.\n2. We see the longer name which is most common in State.\n'

# Bonus Question

1: How can you use the .sample() method to randomly select a single row from the babynames DataFrame, and what is the purpose of such a selection?

2: Suppose you want to generate a random sample of 10 baby names from the babynames DataFrame. How would you use the .sample() method to achieve this, and what parameters or arguments would you specify?

3: You want to perform a bootstrap resampling analysis on the baby names data in the babynames DataFrame. Describe how you can repeatedly sample with replacement using .sample() to create multiple bootstrap samples for statistical analysis

In [17]:
#Part 1
babynames.sample(1,replace=True)
#Use .sample() and explain why should/could we use it to display a single row
''' We use .sample() method becausew it gives us a random row from the data,
The reason behind that Randomly selecting a row can be useful 
for quickly examining a random entry, checking data quality, or getting an example name without bias.  '''
#Part 2
babynames.sample(10,replace=True)
#Use .sample() and explain why should/could we use it to display a ten rows.
''' This can be helpful for quickly exploring the diversity of names in the dataset.
By examining multiple entries, you can identify trends, patterns, that 
may not be visible when looking at just one row.'''
#Part 3

# Assuming you have the 'babynames' DataFrame

# Set the random seed for reproducibility (read reproducibility in documentation and check out what random_seed does)

# Number of bootstrap samples

# Sample size (same as the original dataset)

# Perform bootstrapping

    # Generate a bootstrap sample with replacement

    # Perform your statistical analysis on the bootstrap sample here
    # For example, you can calculate the mean, median, or other statistics

    # Append the statistic of interest to the list

#Hint: The below link is fun to explore and figure out in more detail what all this is:
#https://stackoverflow.com/questions/31502958/drawing-a-bootstrap-sample-from-a-pandas-dataframe

' This can be helpful for quickly exploring the diversity of names in the dataset.\nBy examining multiple entries, you can identify trends, patterns, that \nmay not be visible when looking at just one row.'